homeassistant/components/toon/__init__.py

```diff
@@ -1,142 +1,195 @@
 """Support for Toon van Eneco devices."""
-from datetime import datetime, timedelta
 import logging
+from typing import Any, Dict
 
 import voluptuous as vol
 
 from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
-import homeassistant.helpers.config_validation as cv
-from homeassistant.helpers.discovery import load_platform
-from homeassistant.util import Throttle
+from homeassistant.helpers import (config_validation as cv,
+                                   device_registry as dr)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
 
-REQUIREMENTS = ['toonlib==1.1.3']
+from . import config_flow  # noqa  pylint_disable=unused-import
+from .const import (
+    CONF_CLIENT_ID, CONF_CLIENT_SECRET, CONF_DISPLAY, CONF_TENANT,
+    DATA_TOON_CLIENT, DATA_TOON_CONFIG, DOMAIN)
 
-_LOGGER = logging.getLogger(__name__)
-
-CONF_GAS = 'gas'
-CONF_SOLAR = 'solar'
-
-DEFAULT_GAS = True
-DEFAULT_SOLAR = False
-DOMAIN = 'toon'
-
-MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=5)
+REQUIREMENTS = ['toonapilib==3.0.9']
 
-TOON_HANDLE = 'toon_handle'
+_LOGGER = logging.getLogger(__name__)
 
 # Validation of the user's configuration
 CONFIG_SCHEMA = vol.Schema({
     DOMAIN: vol.Schema({
-        vol.Required(CONF_USERNAME): cv.string,
-        vol.Required(CONF_PASSWORD): cv.string,
-        vol.Optional(CONF_GAS, default=DEFAULT_GAS): cv.boolean,
-        vol.Optional(CONF_SOLAR, default=DEFAULT_SOLAR): cv.boolean,
+        vol.Required(CONF_CLIENT_ID): cv.string,
+        vol.Required(CONF_CLIENT_SECRET): cv.string,
     }),
 }, extra=vol.ALLOW_EXTRA)
 
 
-def setup(hass, config):
-    """Set up the Toon component."""
-    from toonlib import InvalidCredentials
-    gas = config[DOMAIN][CONF_GAS]
-    solar = config[DOMAIN][CONF_SOLAR]
-    username = config[DOMAIN][CONF_USERNAME]
-    password = config[DOMAIN][CONF_PASSWORD]
+async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
+    """Set up the Toon components."""
+    if DOMAIN not in config:
+        return True
 
-    try:
-        hass.data[TOON_HANDLE] = ToonDataStore(username, password, gas, solar)
-    except InvalidCredentials:
-        return False
+    conf = config[DOMAIN]
 
-    for platform in ('climate', 'sensor', 'switch'):
-        load_platform(hass, platform, DOMAIN, {}, config)
+    # Store config to be used during entry setup
+    hass.data[DATA_TOON_CONFIG] = conf
 
     return True
 
 
-class ToonDataStore:
-    """An object to store the Toon data."""
+async def async_setup_entry(hass: HomeAssistantType,
+                            entry: ConfigType) -> bool:
+    """Set up Toon from a config entry."""
+    from toonapilib import Toon
+
+    conf = hass.data.get(DATA_TOON_CONFIG)
+
+    toon = Toon(entry.data[CONF_USERNAME], entry.data[CONF_PASSWORD],
```

Comment:
This seems to make requests upon instantiation, ie I/O. We're not allowed to do I/O in coroutines, as they are executed in the event loop.